Automatic/configurable spektrum bind facility

jflyper opened the pull request:
PR status: Need testing

Based on discussions in #3294 and #1766, this is a second (or third if we count #1766) cut of the series.

- Spektrum bind support and bind plug support is now declared explicitly as `USE_SPEKTRUM_BIND` and `USE_SPEKTRUM_BIND_PLUG` respectively.

- Bind pin defaults to RX or TX pin of the designated serial port.
The code to do this is embedded inside `spektrumBind` for now.
Half-duplex case is handled (with a copy of relevant code from `spektrumInit`), but not tested.
- `SPEKTRUM_BIND_PIN` can be used to override the default.
- Resource `RX_BIND` can also be used to override the default.

If `USE_SPEKTRUM_BIND_PLUG` is declared,
- Bind plug defaults to NONE.
- `BINDPLUG_PIN` can be used to override the default.
- Resource `RX_BIND_PLUG` can also be used to override the default.

AndersHoglund commented:
Great stuff! 
Did a test on my REVO, by removing the hard coded `SPEKTRUM_BIND_PIN` definition and added the new `USE_SPEKTRUM_BIND` in REVO/target.h. Now binding follows port selection. Tested changing serial rx between UART 3 and 6. UART1 of course does not work due to the inverter. Overriding and locking to a specific pin with for example `resource RX_BIND B11` (UART3  Rx) also seems to work.

However, when moving the Sat data line to the UART Tx pin and setting `set serialrx_provider = SRXL` no binding is initiated. Hmm, will do some more checking. But not right now...

mikeller commented on src/main/target/OMNIBUS/target.h:
```diff
@@ -169,9 +169,10 @@
-#define SPEKTRUM_BIND_PIN       UART3_RX_PIN
-
-#define BINDPLUG_PIN            PB0
+#define USE_SPEKTRUM_BIND
```
I think with this implementation, `USE_SPEKTRUM_BIND` should be enabled for all targets that have enough flash space and support Spektrum RX.

jflyper replied:
+1
I'll do it once the code is stable.

mikeller commented on src/main/target/OMNIBUS/target.h:
```diff
@@ -169,9 +169,10 @@
-#define BINDPLUG_PIN            PB0
+#define USE_SPEKTRUM_BIND
+#define USE_SPEKTRUM_BIND_PLUG
+//#define SPEKTRUM_BIND_PIN       NONE // Place holder
```
Not needed any more.

jflyper replied:
Yeah, just wanted to remind my self these are the options ;)

jflyper commented:
@AndersHoglund SRXL now works (under the scope)!

jflyper commented:
@mikeller 
> I think with this implementation, USE_SPEKTRUM_BIND should be enabled for all targets that have enough flash space and support Spektrum RX.

+1

I'll do it once the code is stable.

AndersHoglund commented:
Greatness! 
Binding now works with SRXL on UART1, UART3 and UART6 with Reciever data line connected to UART Tx pin.  Yes, even UART1 works, as we are bypassing the inverter on the UART Rx line.
Sorry for the late response, flying weather today. All day at the field.

AndersHoglund commented on src/main/rx/spektrum.c:
```diff
@@ -210,52 +207,80 @@ bool spekShouldBind(uint8_t spektrum_sat_bind)
+
+        switch (rxConfig->serialrx_provider) {
+        case SERIALRX_SRXL:
+#ifdef TELEMETRY
```
Is it really needed to have a compile condition on TELEMETRY ? And also the next line a runtime check for the TELEMETRY feature being turned on? Binding has no dependency to TELEMETRY as such. Of course you have to turn on TELEMETRY for SRXL to work fully and return telemetry data, but that has nothing to do with binding. 
Unnecessary conditions and constraints might/will lead to support questions and issues.

mikeller replied:
I think this is needed to correctly select the pin the bind pulses need to be sent on, which has to match what is done when opening the serial port:

https://github.com/jflyper/cleanflight/blob/3d2fe5e8de060208a328025a7972b9f38ac2bf46/src/main/rx/spektrum.c#L332

jflyper replied:
I think the conditional on TELEMETRY can be removed.

The code section was stolen from `spektrumInit`, and I believe the conditional is there to save flash.

jflyper commented:
@all Shall we enable `USE_SPEKTRUM_BIND` and `USE_SPEKTRUM_BIND_PIN` for all relevant targets?

AndersHoglund commented:
Are there any target NOT supporting Spektrum ? If not, it could all be moved to target/common_pre.h

jflyper commented:
@AndersHoglund Working on it !!!

jflyper commented:
@kc10kevin Can you take a look at target.h for FURYF7? I don't know the semantics behind PB11 for bind pin.

jflyper commented:
Fixed `isMPUSoftReset` for F4 and F7.

Also added bind pin resetting to input floating when pin is overridden to avoid the pin to interfere with an actual rx pin. This needs some testing with someone with actual receiver and an FC with separate bind pin... ASGARD, or OMNIBUS series with both SBUS and PPM jumper closed.

jflyper commented:
@NightHawk32 Your PR has evolved into this.

mikeller commented:
@jflyper: Is this good to go?

jflyper commented:
@mikeller Let's go now.
